Cython/Compiler/ExprNodes.py

```diff
@@ -4609,8 +4609,11 @@ def generate_result_code(self, code):
             # is_temp is True, so must pull out value and incref it.
             # NOTE: object temporary results for nodes are declared
             #       as PyObject *, so we need a cast
-            code.putln("%s = (PyObject *) *%s;" % (self.result(), self.buffer_ptr_code))
-            code.putln("__Pyx_INCREF((PyObject*)%s);" % self.result())
+            res = self.result()
+            code.putln("%s = (PyObject *) *%s;" % (res, self.buffer_ptr_code))
+            # NumPy does (occasionally) allow NULL to denote None.
+            code.putln("if (%s == NULL) %s = Py_None;" % (res, res))
```

Comment:
I think we should declare this case as unlikely to the C compiler, just in case it can make something of it.
```suggestion
            code.putln("if (unlikely(%s == NULL)) %s = Py_None;" % (res, res))
```